### PR TITLE
INT-3975: AMQP Channels: Support extract-payload

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3975

Also, switch the default mapping to map all inbound headers, but
don't map `x-*` on outbound (by default).

This will provide apps access to important headers such as `x-death` by
default, while not propagating dangerous headers outbound.

### CODE REVIEW

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/channel/PollableAmqpChannel.java [hunk 1]
```diff
@@ -54,12 +56,35 @@
 
 	private volatile int executorInterceptorsSize;
 
+	/**
+	 * Construct an instance with the suppleid name, template and defaule header mappers
```
Comment (artembilan) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/channel/PollableAmqpChannel.java]: Typos: `supplied` and `default`

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AmqpOutboundEndpoint.java [hunk 2]
```diff
@@ -127,7 +128,8 @@ private void send(String exchangeName, String routingKey,
 			final Message<?> requestMessage, CorrelationData correlationData) {
 		if (this.amqpTemplate instanceof RabbitTemplate) {
 			MessageConverter converter = ((RabbitTemplate) this.amqpTemplate).getMessageConverter();
-			org.springframework.amqp.core.Message amqpMessage = mapMessage(requestMessage, converter);
+			org.springframework.amqp.core.Message amqpMessage = MappingUtils.mapMessage(requestMessage, converter,
+					getHeaderMapper(), this.getDefaultDeliveryMode());
```
Comment (artembilan) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AmqpOutboundEndpoint.java]: `this.` is an overhead for methods.
At least when we modify the code :smile:

Comment (garyrussell) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AmqpOutboundEndpoint.java]: Need a checkstyle :smile:

Comment (artembilan) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AmqpOutboundEndpoint.java]: M-m-m. `RedundantThis` and so on.

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java [hunk 3]
```diff
@@ -429,31 +429,31 @@ public static DefaultAmqpHeaderMapper outboundMapper() {
 	 * @return the default request headers for an inbound mapper.
 	 */
 	public static String[] inboundRequestHeaders() {
-		return safeInboundHeaders();
+		return new String[] { "*" };
```
Comment (artembilan) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java]: Not related to the changes, but I wonder why we don't see any deprecation warn for the factory methods based on the `@Deprecated` ctor:
```
public static DefaultAmqpHeaderMapper outboundMapper() {
	DefaultAmqpHeaderMapper mapper = new DefaultAmqpHeaderMapper();
	...
}
```
JDK bug?

We should provide `private` ctor on the matter and don't use deprecated API from our code.
Isn't it?

**Comment by artembilan** [comment]
Gary, that's all.
I don't see any other issues and can address all those minor concerns on merge.

**Comment by garyrussell** [comment]
Pushed

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java [hunk 4]
```diff
@@ -411,22 +421,33 @@ private void addConsumerMetadata(MessageProperties messageProperties, Map<String
...
+	 * @see #outboundRequestHeaders()
+	 * @see #outboundReplyHeaders()
+	 * @since 4.3
+	 */
 	public static DefaultAmqpHeaderMapper outboundMapper() {
-		DefaultAmqpHeaderMapper mapper = new DefaultAmqpHeaderMapper();
-		mapper.setRequestHeaderNames(outboundRequestHeaders());
-		mapper.setReplyHeaderNames(outboundReplyHeaders());
+		DefaultAmqpHeaderMapper mapper = new DefaultAmqpHeaderMapper(outboundRequestHeaders(), outboundReplyHeaders());
 		return mapper;
```
Comment (artembilan) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java]: Redundant variable.
Will fix on merge.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/dc94b420eef6a8d9053a28d2bfb11a0d7e714452

**Pull request closed without merging**